File: Chapter 02/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter02/section_2_3_2_maintainability/before/ShoppingCart.java
package com.grokkingsoftwarearchitecture.chapter02.section_2_3_2_maintainability.before;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;

// ARCHITECTURAL NOTE: The "File Dump" Anti-Pattern
// In messy codebases, developers often dump data models (like CartItem) 
// into the exact same file as the business logic. As the app grows, 
// this file will become thousands of lines long and impossible to navigate.
class CartItem {
    private String name;
    private BigDecimal price;

    public CartItem(String name, BigDecimal price) {
        this.name = name;
        this.price = price;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }
}

/**
 * Manages shopping cart operations.
 * ARCHITECTURAL NOTE: This "Before" class demonstrates poor maintainability 
 * due to magic numbers and a lack of Separation of Concerns.
 */
public class ShoppingCart {
    
    public String processOrder(List<CartItem> cartItems) {
        // 1. Calculating the subtotal.
        BigDecimal subtotal = BigDecimal.ZERO;
        for (CartItem item : cartItems) {
            subtotal = subtotal.add(item.getPrice());
        }

        // PROBLEM 1: "Magic Numbers"
        // The numbers 0.10 and 0.08 are hardcoded values without any explanation.
        BigDecimal discount = subtotal.multiply(new BigDecimal("0.10")); 
        BigDecimal totalAfterDiscount = subtotal.subtract(discount);
        
        BigDecimal tax = totalAfterDiscount.multiply(new BigDecimal("0.08")); 
        BigDecimal finalTotal = totalAfterDiscount.add(tax);

        // PROBLEM 2: Lack of Separation of Concerns
        // This method does everything: calculates subtotal, applies discount, and adds tax.
        return String.format("Order processed! Your final total is $%.2f", finalTotal);
    }
}

File: Chapter 02/C#/Section_2_3_2_Maintainability/Before/ShoppingCart.cs
namespace Chapter02.ShoppingCart.Before;

// ARCHITECTURAL NOTE: The "File Dump" Anti-Pattern
// In messy codebases, developers often dump data models (like CartItem) 
// into the exact same file as the business logic. As the app grows, 
// this file will become thousands of lines long and impossible to navigate.
public class CartItem
{
    public required string Name { get; set; }
    public decimal Price { get; set; }
}

/// <summary>
/// Manages shopping cart operations.
/// ARCHITECTURAL NOTE: This "Before" class demonstrates poor maintainability 
/// due to magic numbers and a lack of Separation of Concerns.
/// </summary>
public class ShoppingCart
{
    public string ProcessOrder(List<CartItem> cartItems)
    {
        // 1. Calculating the subtotal.
        decimal subtotal = 0;
        foreach (var item in cartItems)
        {
            subtotal += item.Price;
        }

        // PROBLEM 1: "Magic Numbers"
        // The numbers 0.10 and 0.08 are hardcoded values without any explanation.
        decimal discount = subtotal * 0.10m; 
        decimal totalAfterDiscount = subtotal - discount;
        
        decimal tax = totalAfterDiscount * 0.08m; 
        decimal finalTotal = totalAfterDiscount + tax;

        // PROBLEM 2: Lack of Separation of Concerns
        // This method does everything: calculates subtotal, applies discount, and adds tax.
        return $"Order processed! Your final total is ${finalTotal:F2}";
    }
}

public class Demo
{
    public static void Run()
    {
        Console.WriteLine("--- Maintainability: Shopping Cart (BEFORE) ---");
        Console.WriteLine("Notice the 'magic numbers' and the rigid 'God Method' design.\n");
        
        var cart = new List<CartItem>
        {
            new CartItem { Name = "Laptop", Price = 1000.00m },
            new CartItem { Name = "Mouse", Price = 50.00m }
        };

        var cartSystem = new ShoppingCart();
        Console.WriteLine(cartSystem.ProcessOrder(cart));
        Console.WriteLine("\n-----------------------------------------");
    }
}